whylog/teacher/constraint_links_base.py

```diff
@@ -0,0 +1,51 @@
+class ConstraintLinksBase:
+    """
+    Container for links between group and constraint
+    """
+
+    LINE_ID_POSITION = 0
+    GROUP_NUMBER_POSITION = 1
+    CONSTRAINT_ID_POSITION = 2
+
+    def __init__(self, links_list=None):
+        """
+        :param links_list: list of links between group and constraint
+                          one link is a tuple (line_id, group_no, constraint_id)
+        :type links_list: list[(int, int, int)]
+        """
+        self.links = links_list or []
+
+    def get_links(self):
+        return self.links
+
+    def distinct_constraint_ids(self):
+        return set(list(zip(*self.links))[self.CONSTRAINT_ID_POSITION])
+
+    def add_links(self, links_list):
+        self.links = list(set(self.links + links_list))
```

Comment:
`self.links = list(set(self.links) | links_list))` or something, you should use the set operation for this in case links_list is an iterable, and not an actual list. You should not put a "list" in the interface, but an iterable instead, as practically all functions do.